### ISSUE TITLE
How to refresh ECR credentials in a non-AWS Rancher 2.x installation

### ISSUE THREAD

**Issue by shuckepco** [rancher/rancher#14907]
**Rancher versions:**
Rancher v2.0.6 - installed via RKE 0.1.8

**Infrastructure Stack versions:**
Kubelet Version 1.10.5
Kube Proxy Version 1.10.5

**Docker version: (`docker version`,`docker info` preferred)**
 Version:      17.03.2-ce

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
HA rancher (running rancher in the K8s cluster)

_Stripped the parts that seemed to be definitely irrelevant for my question..._

**My question:**
We just installed Rancher v2.x on a 3-node cluster in our own datacenter. We store our docker images in a private Amazon ECR registry. We also have a Rancher v1.x installation where we use the ecr-updater from your catalog to continuously update the credentials. Unfortunately, we have found no information on how to achieve this in v2. **How can we use the ECR registry in a local Rancher v2.x installation? And how can Rancher/K8s refresh the credentials automatically?**

Thanks in advance!

**Comment by sehucke** [comment]
For whom it may concern:
We have managed that by using https://github.com/trondhindenes/ecr-updater (with some minor tweaks)

But there is a small bug (see https://github.com/rancher/rancher/issues/13339) that needs to be worked around until fixed.

**Comment by loganhz** [comment]
Please try https://github.com/rancher/rancher/issues/14907#issuecomment-418864667

**Comment by loganhz** [comment]
Maybe we should package the `ecr-updater` into a catalog app.

**Comment by omniscient** [comment]
I need help connecting the dots here. I tried using https://github.com/trondhindenes/ecr-updater but it's not working for me: 

> container_linux.go:247: starting container process caused "process_linux.go:295: setting oom score for ready process caused \"write /proc/18910/oom_score_adj: invalid argument\""

Can someone describe in a bit more words what has to be done to get ECR working with Rancher 2.0.x?

**Comment by flybd5** [comment]
The easiest way to create and update a docker-registry secret for an AWS ECR repo is to set up a CRON job that uses the AWS CLI and Kubectl.

- Log into the host machine where the cluster is running
- Install AWS CLI and configure it to use an IAM role that can read the ECR credentials.
- Test the CLI config using this command: `aws ecr --region <your_ecr's_region> get-login-password`
- Install kubectl and configure it with the yaml for the cluster in which you want to define the secret
- Create a shell script like this:

```
# Delete the secret if it already exists (there is no way to update it)
kubectl delete secret <name_of_the_docker_registry_secret>
# Create the secret
kubectl create secret docker-registry <name_of_the_docker_registry_secret> \
	--docker-server=<your_ecr_registry_uri> \
	--docker-username=AWS \
	--docker-password=$(aws ecr --region <your_ecr's_region> get-login-password) \
	--docker-email=<your_notification_email>
```

- Now setup a cron job to run this shell script every 6 hours (AWS resets the password every 12, so just to be on the safe side).


The secret will show up in the default project for the cluster. If you know how to have kubectl create it in a specific project, let me know.

**Comment by florissmit10** [comment]
Hello,
We are going to use rancher to manage a fleet of single node k3s clusters. They are distributed on premise, and cannot use the default aws ecr credentials route.
 It would be nice to have some tool to sync these credentials from our central aws based rancher setup. 
Any plans to add some functionality for this?

**Comment by flybd5** [comment]
No, I don't have access anymore to the AWS infra I used to develop this. Also, credentials don't come from Rancher, they come from AWS. If you are not using ECR then this technique is not applicable.
